pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,140 @@
+	Assign(resource interface{}, value interface{}) error
+}
+
+// ResolveReferencers resolves references made to other resources
+func ResolveReferencers(ctx context.Context, mgd Managed, c client.Reader) (bool, error) {
+	resolvers, err := findRefResolverFields(mgd, false)
+	if err != nil {
+		return false, err
+	}
+
+	allRferencesReady, err := checkIfAllReferencesAreReady(getAllObjectReferences(resolvers))
+	if err != nil || !allRferencesReady {
+		return false, err
+	}
+
+	for _, r := range resolvers {
+
+		val, err := r.Resolve(c)
+		if err != nil {
+			return allRferencesReady, err
+		}
+
+		err = r.Assign(mgd, val)
+		if err != nil {
+			return allRferencesReady, err
+		}
+	}
+
+	return false, nil
+}
+
+// findRefResolverFields recursively finds all non-nil fields in a struct and its sub-structs
+// that implement ReferenceResolver and have `managed_resource:"reference_resolver"` tag
+func findRefResolverFields(obj interface{}, objHasTheRightTag bool) ([]ReferenceResolver, error) { // nolint:gocyclo
+
+	v := reflect.ValueOf(obj)
+	result := []ReferenceResolver{}
+
+	// ignore nil fields
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return nil, nil
+	}
+
+	if objHasTheRightTag {
+		if ar, implements := obj.(ReferenceResolver); implements {
+			return []ReferenceResolver{ar}, nil
+		} else if v.Kind() == reflect.Struct {
+			return nil, fmt.Errorf("the field has the 'reference_resolver' tag, but has not implemented ReferenceResolver interface")
+		}
+	}
+
+	switch v.Kind() {
+	case reflect.Struct:
+		for i := 0; i < v.NumField() && v.Field(i).CanInterface(); i++ {
+			val, ok := reflect.TypeOf(obj).Field(i).Tag.Lookup("managed_resource")
+			hasTheRightTag := ok && val == "reference_resolver"
+
+			if resolvers, err := findRefResolverFields(v.Field(i).Interface(), hasTheRightTag); err == nil {
+				result = append(result, resolvers...)
+			} else {
+				return nil, err
+			}
+		}
+
+	case reflect.Slice:
+		for i := 0; i < v.Len() && v.Index(i).CanInterface(); i++ {
+			if resolvers, err := findRefResolverFields(v.Index(i).Interface(), objHasTheRightTag); err == nil {
+				result = append(result, resolvers...)
+			} else {
+				return nil, err
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// getAllObjectReferences retrieves all the fieds that or of type `corev1.LocalObjectReference`
+// from the given interfaces that implemente ReferenceResoler
+func getAllObjectReferences(resolvers []ReferenceResolver) []corev1.LocalObjectReference {
+	result := []corev1.LocalObjectReference{}
+
+	for _, resolver := range resolvers {
+		r := reflect.ValueOf(resolver).Elem()
+		for i := 0; i < r.NumField() && r.Field(i).CanInterface(); i++ {
+			if ref, ok := r.Field(i).Interface().(corev1.LocalObjectReference); ok {
+				result = append(result, ref)
+			}
+
+			if ref, ok := r.Field(i).Interface().(*corev1.LocalObjectReference); ok {
+				result = append(result, *ref)
+			}
```

Comment:
Is the intent here to support both required and optional references (i.e. object reference and pointer to object reference)? Could we simplify this by requiring that all references are either required or optional?

Reply:
I think that's a good idea.